SPARK-4585. Spark dynamic executor allocation should use minExecutors as...

sryza opened the pull request:
... initial number

sryza committed:
SPARK-4585. Spark dynamic executor allocation should use minExecutors as initial number

SparkQA commented:
[Test build #25568 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25568/consoleFull) for   PR 4051 at commit [`9d7b6f9`](https://github.com/apache/spark/commit/9d7b6f98caff5de3db88d372853cccf012f36dc6).
 * This patch merges cleanly.

ksakellis commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/ClientArguments.scala:
```diff
@@ -73,12 +73,12 @@ private[spark] class ClientArguments(args: Array[String], sparkConf: SparkConf)
       .orNull
     // If dynamic allocation is enabled, start at the max number of executors
```
Fix

ksakellis commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/ClientArguments.scala:
```diff
@@ -73,12 +73,12 @@ private[spark] class ClientArguments(args: Array[String], sparkConf: SparkConf)
-      val maxExecutorsConf = "spark.dynamicAllocation.maxExecutors"
-      if (!sparkConf.contains(maxExecutorsConf)) {
+      val minExecutorsConf = "spark.dynamicAllocation.maxExecutors"
+      if (!sparkConf.contains(minExecutorsConf)) {
```
so the ExecutorAllocationManager already validates that minExecutorConf exists and is valid (not negative and non-zero). Do we need this? or are we being paranoid? If so, maybe an assertion is better?

sryza replied:
I think the rationale may have been that the `ExecutorAllocationManager` and client can run in different processes.

SparkQA commented:
[Test build #25568 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25568/consoleFull) for   PR 4051 at commit [`9d7b6f9`](https://github.com/apache/spark/commit/9d7b6f98caff5de3db88d372853cccf012f36dc6).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25568/
Test PASSed.

andrewor14 commented:
Hi @sryza I have read through the discussion on the JIRA. The original motivation for making it the max is that it is common for an application to start running jobs as soon as they are launched. Setting it to the min, however, will add some initial application startup delay for most cases. Isn't better to try to acquire many resources and give them back to the cluster as needed?

sryza commented:
To distill the motivation on the JIRA and make sure we're on the same page: in most situations (including Hive-on-Spark), users don't or can't know how many resources their app will be able to take advantage of, so they want the max to be effectively unbounded.  Requesting all these resources on startup wouldn't make sense.

If the concern is that there will be application startup delay, maybe it would make sense to add an initialExecutors property?
 

I think in most situations

andrewor14 commented:
I see, maybe that makes more sense. The natural follow up question is whether to default that to the min or the max. The argument for the min is exactly what you just described, and the argument for the max is that the behavior will be consistent with 1.2 if the users don't set this new config.

ksakellis commented:
I'd prefer we don't introduce yet another config option for dynamic scaling. How much delay are we really introducing here? The number of running executors will get to maxExecutors (if there are more pending tasks) within one yarn call. Is the single yarn call really that slow?

pwendell commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/ClientArguments.scala:
```diff
@@ -73,12 +73,12 @@ private[spark] class ClientArguments(args: Array[String], sparkConf: SparkConf)
     if (isDynamicAllocationEnabled) {
-      val maxExecutorsConf = "spark.dynamicAllocation.maxExecutors"
-      if (!sparkConf.contains(maxExecutorsConf)) {
+      val minExecutorsConf = "spark.dynamicAllocation.maxExecutors"
```
Should this be "spark.dynamicAllcoation.minExecutors" on the right side?

pwendell commented:
It seems fine to me to just change the defaults. This is an experimental feature in 1.2 and people will just be kicking the tires. If we get some user feedback on this later, we can always add a configuration to preserve the legacy behavior.

lianhuiwang commented:
when we set initial number to min, there is a delay to request more executors. because in ExecutorAllocationManager's addExecutors numExecutorsToAdd starts from 1, and next minute is double. so when we request more executors, there is very slow and it spends several minutes.so maybe we need to change this scaling mechanism? 
another situation is:
when stage has skew task and  individual tasks need to run for a long time, number of executor will be to a small value, and then next stage need many executors to run tasks, so ExecutorAllocationManager need to request many executors. at this time it also need to spend a few minutes for reaching number of pending tasks. @andrewor14

sryza committed:
Add spark.dynamicAllocation.initialNumExecutors and make min and max not required

sryza commented:
I updated the patch to add a `spark.dynamicAllocation.initialExecutors` property. I also removed the requirement to set min/maxExecutors, so the user now only needs to configure any numbers if they want to tune for performance.  The min now defaults to 0 and the max to Integer.MAX_VALUE.  I looked through the code and it seems like this should be safe, but @andrewor14 let me know if there are reasons I missed why these needed to be required.

SparkQA commented:
[Test build #25802 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25802/consoleFull) for   PR 4051 at commit [`d1dd039`](https://github.com/apache/spark/commit/d1dd039c1807cab11f998b76b698d19e2565ec53).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25802 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25802/consoleFull) for   PR 4051 at commit [`d1dd039`](https://github.com/apache/spark/commit/d1dd039c1807cab11f998b76b698d19e2565ec53).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25802/
Test PASSed.

lianhuiwang commented:
@sryza @andrewor14 i find that setting minExecutors to initialExecutors  is best for the following situation: when DAGScheduler submits missing tasks of first stage very lately because DAGScheduler  need to scan large hdfs path in order to get rdd's partitions. in my test, some jobs cost more than two minutes on scanning hdfs. so in this situation, if we set a large number to initialExecutors, before DAGScheduler submit missing tasks, these resources of initialExecutors were wasted. if the time > executorIdleTimeout, initialExecutors  will be removed since they are executorIdleTimeout.
and we can reduce schedulerBacklogTimeout to make numExecutors reach to maxExecutors quickly.

andrewor14 commented:
@sryza Is the point of not requiring these configs that the users don't really know how many executors they actually want? I'm OK with not requiring a min (and defaulting it to at least 1, but not 0), but in YARN, even without dynamic allocation users already have to specify the number of executors, so I think it's fine to require the user to set the max. Also defaulting the max to Int.MaxValue doesn't make much sense in my opinion because the whole point of doing this is to share resources in the cluster.

@lianhuiwang @ksakellis The initial scale up delay is the reason why I originally set the starting number to the max. I understand that introducing an additional config may be confusing since we already have `--num-executors` and `spark.executor.instances`. We have two options here: (1) introduce a `spark.dynamicAllocation.initialNumExecutors` or something, or (2) use the existing setting of `--num-executors` or `spark.executor.instances` as the starting point. I personally prefer (2) even though it might confuse existing users because in Spark 1.2.0 we actually *disallow* users from setting `--num-executors` if they're using dynamic allocation. I think that might be the best option here though.

lianhuiwang commented:
@andrewor14 i think replacing maxExecutors  with --num-executors is more reasonable, because when dynamic allocation is not enable, --num-executors is the maxExecutors. for some of running jobs,  --num-executors is more large than initialNumExecutors, so i think initialNumExecutors is better than --num-executors.
and i agree with that we need maxExecutors to avoid that a job can get Int.MaxValue resources.

sryza commented:
> @sryza Is the point of not requiring these configs that the users don't really know how many executors they actually want?

Exactly.  From my perspective, one of the main motivations for dynamic allocation in the first place was removing the need to set `--num-executors`.  People have long complained about the long list configurations users need to know for MapReduce, but at this point Spark actually requires more to get decent resource utilization.  This particular JIRA was filed Hive didn't know how to set the max executors in a general way.  What specific advice would you give to a user about how to set this property?

> Also defaulting the max to Int.MaxValue doesn't make much sense in my opinion because the whole point of doing this is to share resources in the cluster.

This is the job of the cluster resource manager.  Users and apps shouldn't need self-regulate their usage. YARN will throttle resources when there is contention.  The main responsibility of the app is just to not ask for / sit on more resources than it can take advantage of, which dynamic allocation helps with in general, and which SPARK-4136 will help enforce more strictly.

vanzin commented:
A few comments:
- I don't have a strong opinion on new option vs. not. I think eventually what we really want is to always have dynamic allocation on, so that discussion would eventually become moot.
- min = 0 makes sense; this is particularly interesting for things like spark-shell, or Hive sessions. As long as ramp up is reasonably quick when needed, it should be fine. (0 vs. 1 probably won't make a big difference for large jobs anyway.)
- I'm not so sure about `Integer.MAX_VALUE`. I get the point of letting the resource manager handle it, but perhaps we should be nicer here? e.g., set max to "number of NMs in the cluster" if that info is available client-side?

sryza commented:
What would be the advantage of limiting executor requests to the number of NMs in the cluster?

vanzin commented:
It feels like it would be better to have "# of NMs" executors with many cores than "> # of NMs" executors with less cores.

Unless your plan is that we should prefer lots of small executors launched on demand, in lieu of the current approach people take of launching larger executors.

ksakellis commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/ClientArguments.scala:
```diff
@@ -75,14 +75,23 @@ private[spark] class ClientArguments(args: Array[String], sparkConf: SparkConf)
+      val initialExecutorsConf = "spark.dynamicAllocation.initialExecutors"
       val maxExecutorsConf = "spark.dynamicAllocation.maxExecutors"
-      if (!sparkConf.contains(maxExecutorsConf)) {
+      val minNumExecutors = sparkConf.getInt(minExecutorsConf, 0)
```
Nit: indentation under if statement

ksakellis commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/ClientArguments.scala:
```diff
@@ -75,14 +75,23 @@ private[spark] class ClientArguments(args: Array[String], sparkConf: SparkConf)
+      val maxNumExecutors = sparkConf.getInt(maxExecutorsConf, Integer.MAX_VALUE)
+
+      // If defined, initial executors must be between min and max
+      if (initialNumExecutors < minNumExecutors || initialNumExecutors > maxNumExecutors) {
```
Can you add this check in ExecutorAllocationManager where the other validations exist?

ksakellis commented:
LGTM

I agree with @sryza that maxExecutors should default to Integer.MAX_VALUE. It is not the app's responsibility to play fair but rather to ask the resource manager, YARN, for its optimal amount of resources and it take care of enforcing the admin policies. @vanzin I also don't see the point with the app setting the max to #nm. If you default maxExecutors to #nm on startup, what happens when the administrator adds more node managers while the spark job is running? It would be good for Spark to take advantage of the spare capacity. 

Although I don't like yet another configuration option, initialExecutors, I don't think there is a way around it given the hive on spark use case. At least none of the three are required which is the most important part. A default value of 0 for minExecutors seems like a good default for the reasons described by @vanzin. Defaulting initialExecutors to minExecutors seems fine also. @andrewor14 I don't like the idea of using --num-executors as the initialExecutors. Like @vanzin we want at some point to make dynamic allocation the default so having consistent configuration make sense. Reusing --num-executors would just add confusion now and later.

@andrewor14 why do we increase the number of executors we request once a minute? is this to reduce load on the YARN resource manager? I think we might want to revisit how we ramp up the cluster but that is beyond the scope of this PR.

vanzin commented:
> I also don't see the point with the app setting the max to #nm.

I was worried that effectively having no upper bound would make the allocator just queue up lots and lots of requests that would never be serviced. But Sandy told me offline that the allocator won't request more containers than the number of pending tasks require, so in that case it should be fine.

andrewor14 commented:
Alright, in summary, the proposal here is to default the min to 0 and the max to Int.Max, and add a new `initialExecutors` config that defaults to the min. On second thought I think the max sounds reasonable because the number executors we get is actually limited by the number of pending tasks we get (so it'll never actually ramp up all the way). As for the min, if we do decide to set it to 0 then perhaps we should lower the ramp up interval, though we can do that separately. I just thought the advantage of keeping it at 1 is that we'll always make progress no matter how slowly we ramp it up, though it's not a big enough issue to block this patch.

I'm merging this into master. Thanks for the input everyone.

pwendell commented:
For some reason this didn't close. @sryza can you close this issue?
